### PR TITLE
Snowflake: Add Create Storage Integration grammar.

### PR DESCRIPTION
**Pull request by jmc-bbk**
### Brief summary of the change made

We don't currently support `CREATE ... STORAGE INTEGRATION` grammar in Snowflake and as a result this is unparsable.

https://docs.snowflake.com/en/sql-reference/sql/create-storage-integration.html

Added CREATE Storage Integration grammar to Snowflake.

### Are there any other side effects of this change that we should be aware of?

Not that I'm aware of. 

### Pull Request checklist
- [x] Please confirm you have completed any of the necessary steps below.

- Included test cases to demonstrate any code changes, which may be one or more of the following:
  - `.yml` rule test cases in `test/fixtures/rules/std_rule_cases`.
  - `.sql`/`.yml` parser test cases in `test/fixtures/dialects` (note YML files can be auto generated with `tox -e generate-fixture-yml`).
  - Full autofix test cases in `test/fixtures/linter/autofix`.
  - Other.
- Added appropriate documentation for the change.
- Created GitHub issues for any relevant followup/future enhancements if appropriate.

### CODE REVIEW

**Comment by jmc-bbk** [comment]
I have left a couple of `TODO:` statements that it would be great to receive some help with.

- We should limit one argument to be `'bucket-owner-full-control'` which is the only valid string instead of any string.
- One argument should be a quoted wildcard character (`'*'`). I'm unsure how to handle this - similar to above.
- Finally, I left lots of comments (which can be removed) about the valid PATHs for `STORAGE_ALLOWED_LOCATIONS`. I split out S3, GCS, and Azure into their own grammars, as they have their own unique parameters. However, the current set up would (e.g.) allow you to have S3 cloud parameters with GCS paths in `STORAGE_ALLOWED_LOCATIONS`.

On the last point above - I could move `STORAGE_ALLOWED_LOCATIONS` and `STORAGE_BLOCKED_LOCATIONS` into the <Cloud-Provider>StorageIntegrationParameters segments. But I feel like this is less readable/easy to understand when comparing against the Snowflake docs.

Open to opinions and I will either delete my comments in the PR or move the logic 👍 .

File: src/sqlfluff/dialects/dialect_snowflake.py
Comment (tunetheweb) [inline src/sqlfluff/dialects/dialect_snowflake.py:2372]: As this one is used elsewhere, you should add a `QuotedStarSegment` in the section starting at line 89.

```py
    QuotedStarSegment=StringParser(
        "'*'",
        CodeSegment,
        name="quoted_star",
        type="identifier",
        trim_chars=("'",),
    ),
```

Then can change to this:

```suggestion
                                Ref("QuotedStarSegment"),
```

File: src/sqlfluff/dialects/dialect_snowflake.py
Comment (tunetheweb) [inline src/sqlfluff/dialects/dialect_snowflake.py:2389]:
```suggestion
                                Ref("QuotedStarSegment"),
```

File: src/sqlfluff/dialects/dialect_snowflake.py
Comment (tunetheweb) [inline src/sqlfluff/dialects/dialect_snowflake.py:2845]: For one ofs, rather than defining a segment, you can embed a StringParser directly in the dialect:

```suggestion
            StringParser("'bucket-owner-full-control'", CodeSegment, type="literal"),
```

File: src/sqlfluff/dialects/dialect_snowflake.py
Comment (tunetheweb) [inline src/sqlfluff/dialects/dialect_snowflake.py:2360]: It's not necessary to write 100% compliant parsing code. While it would be nice to only parse exactly what Snowflake would accept., you need to weigh up how likely the above is, whether it would be expected for a SQL linter to catch those, versus the complexity of the code you're adding to SQLFluff. We've lots of places where we are overly permissive as to what we allow.

Also how likely is it someone will add S3 in one place and GCS on another? And does Snowflake itself throw a parsing error in this case? Or a runtime error?

Saying that I don't think adding to the IntegrationParameters is that bad. A comment would explain it.

Or alternatively do something like this:

```py
                OneOf(
                    Sequence(
                        Ref("S3StorageIntegrationParameters"),
                        Ref("S3StorageAllowdLocations"),
                    ),
                    Sequence(
                        Ref("GCSStorageIntegrationParameters"),
                        Ref("GCSStorageAllowdLocations"),
                    ),
                    Sequence(
                        Ref("AzureStorageIntegrationParameters"),
                        Ref("AzureStorageAllowdLocations"),
                    ),
                ),
```

**Review by tunetheweb — CHANGES_REQUESTED** [review]

File: src/sqlfluff/dialects/dialect_snowflake.py
Comment (jmc-bbk) [reply src/sqlfluff/dialects/dialect_snowflake.py:2541]: In this case - it would only be a runtime error, but the parsing would be fine.

Given the above, do you think I should leave it as it is (with my comments removed) or I could also change to as you suggest (with slight edit):

```
 OneOf(
                    AnySetOf(
                        Ref("S3StorageIntegrationParameters"),
                        Ref("S3StorageAllowdLocations"),
                    ),
```

I think I'm happy to leave the code as it is - like you mention, we can parse it and so can Snowflake. It's only the equivalent to someone adding a bad S3 path that doesn't actually exist!

Let me know and I can update tomorrow :) Saw your other comments too, thanks super helpful - will implement them tomorrow morn!

Comment (tunetheweb) [reply src/sqlfluff/dialects/dialect_snowflake.py:2541]: Leaving as is SGTM!

**Comment by jmc-bbk** [comment]
Note - I made another minor change that wasn't requested. For both `STORAGE_ALLOWED` and `STORAGE_BLOCKED` I updated this snippet 

```
                        Bracketed(
                            Delimited(
                                OneOf(
                                    Ref("S3Path"),
                                    Ref("GCSPath"),
                                    Ref("AzureBlobStoragePath"),
                                   Ref("QuotedStarSegment"),
                                )
                            )
                        ),
```

to this snippet

```
                    OneOf(
                        Bracketed(
                            Delimited(
                                OneOf(
                                    Ref("S3Path"),
                                    Ref("GCSPath"),
                                    Ref("AzureBlobStoragePath"),
                                )
                            )
                        ),
                        Bracketed(
                            Ref("QuotedStarSegment"),
                        ),
```

Before, I was getting any number of QuotedStarSegments in the `Delimited()` - but you can only have 1 Bracketed, QuotedStarSegment - so I have updated, and tests still pass.

File: src/sqlfluff/dialects/dialect_snowflake.py
Comment (tunetheweb) [inline src/sqlfluff/dialects/dialect_snowflake.py:2580]: As these are very similar I wonder if they can be combined? As they are part of an `AnySetOf` it should still be allowed to have both.

Also the second one ahas an `optional=True` on line 2579 but don't think that's needed as it's part of an `AnySetOf` so optionality is implied.

```suggestion
                Sequence(
                    OneOf("STORAGE_ALLOWED_LOCATIONS", "STORAGE_BLOCKED_LOCATIONS"),
                    Ref("EqualsSegment"),
                    OneOf(
                        Bracketed(
                            Delimited(
                                OneOf(
                                    Ref("S3Path"),
                                    Ref("GCSPath"),
                                    Ref("AzureBlobStoragePath"),
                                )
                            )
                        ),
                        Bracketed(
                            Ref("QuotedStarSegment"),
                        ),
                    ),
                ),
```

File: src/sqlfluff/dialects/dialect_snowflake.py
Comment (tunetheweb) [inline src/sqlfluff/dialects/dialect_snowflake.py:2585]: Don't think this is needed as implied as part of `AnySetOf`

```suggestion
```

**Review by tunetheweb — CHANGES_REQUESTED** [review]
Looks pretty good, but have a few more suggestions and then can merge this.

File: src/sqlfluff/dialects/dialect_snowflake.py
Comment (jmc-bbk) [reply src/sqlfluff/dialects/dialect_snowflake.py:2588]: Ah, I was going from the docs where it has this syntax:

```
...
  ENABLED = { TRUE | FALSE }
  STORAGE_ALLOWED_LOCATIONS = ('<cloud>://<bucket>/<path>/', '<cloud>://<bucket>/<path>/')
  [ STORAGE_BLOCKED_LOCATIONS = ('<cloud>://<bucket>/<path>/', '<cloud>://<bucket>/<path>/') ]
```

so I think it comes down to the parse versus. runtime error issue.

With the above, `STORAGE_ALLOWED_LOCATIONS` is required - but it can come before the other arguments (e.g.) `ENABLED` - whereas `STORAGE_BLOCKED_LOCATIONS` is optional.

I suppose what I'm really looking for is an `UnorderedSequence()` rather than `AnySetOf()` - I didn't actually realise that `AnySetOf()` is implicitly optional.

Is there anything like `UnorderedSequence` available?

If not - I'm happy to do your suggestions of removing the `optional=True` from these. Let me know!

Comment (tunetheweb) [reply src/sqlfluff/dialects/dialect_snowflake.py:2588]: Don't think we can really handle that to be honest. So think we need to basically treat both as optional.

Comment (jmc-bbk) [reply src/sqlfluff/dialects/dialect_snowflake.py:2588]: Sure thing, I'll start to make the above changes :)

**Comment by jmc-bbk** [comment]
Hi @tunetheweb, sorry for the delay here!

I actually noticed from your comments that I had a problem in my original code. `STORAGE_BLOCKED_LOCATIONS` cannot accept `('*')` as a valid argument.

I have now removed that from that section.

Given that makes the `Sequence()` different for `STORAGE_BLOCKED_LOCATIONS` and `STORAGE_ALLOWED_LOCATIONS`, I haven't merged those as in your above suggestion. But I'm still open to, if you think that's the right move?

---

I also looked at the `optional = True` statements you mentioned in the other sections (e.g. `PIPE` / `WAREHOUSE`). I've noticed that all those `Sequence()`s are `optional=True`.

It looks like this `CreateStatementSegment(...)` could be too broad at the moment. It might need to be decomposed into

```
CreatePipeSegment
CreateIntegrationSegment
...
CreateXSegment
```

What do you think?

If the above sounds right, I could work on it in another PR - but maybe it's not in scope of the current work?

Anyway, let me know - and happy to make changes :)

**Comment by tunetheweb** [comment]
I actually typed that as a suggestion previously! 😁 Then delighted it as thought it wasn’t fair to you to have to take that on in this PR. Do agree we should do it as a separate PR once this is merged.

**Comment by jmc-bbk** [comment]
Great! Let me know if there's anything outstanding :)

File: src/sqlfluff/dialects/dialect_snowflake.py
Comment (tunetheweb) [reply src/sqlfluff/dialects/dialect_snowflake.py:2580]: I still don't think we need this `optional=True`

File: src/sqlfluff/dialects/dialect_snowflake.py
Comment (tunetheweb) [inline src/sqlfluff/dialects/dialect_snowflake.py:3022]: Nor this.

```suggestion
```

**Review by tunetheweb — CHANGES_REQUESTED** [review]
Looks pretty much there too me. But still think we can drop two of the `optional=True`s

**Comment by jmc-bbk** [comment]
https://github.com/sqlfluff/sqlfluff/pull/3075#discussion_r850582411

For the optional on `L:2580`, I think we do require it. As I'm setting all the arguments for the `CREATE STORAGE INTEGRATION` syntax to be optional.

If I were to remove that optional, I think we would be unable to parse the next section for `CREATE PIPE` - as it would be expecting the `CREATE STORAGE INTEGRATION` arguments beforehand?

Or I might still be misunderstanding `AnySetOf` 🥴 - I have pushed the change to `L:3022` will try the test suite on my local if I remove the one on `L:2580`.

**Comment by jmc-bbk** [comment]
Just tested `pytest -s test/dialects/dialects_test.py -k snowflake-snowflake_create_pipe.sql` without the `optional=True` on `L:2580` and it would fail!

**Comment by tunetheweb** [comment]
Oh I guess I’m misunderstanding it! I thought `AnySetOf` was 0 or 1 of each of the options. Therefore the 0 was the same as optional.

Maybe you need an optional on the overall `AnySetOf` if none of them apply? And maybe that’s currently being effectively picked up by the fact one of your options is optional?

**Comment by jmc-bbk** [comment]
I'm actually having trouble viewing this in the Github UI - maybe this is causing some confusion - but I think I already have the above suggestion. Hopefully easier to view if I paste the relevant snippet:

```
        Sequence(
            AnySetOf(
                Sequence("TYPE", Ref("EqualsSegment"), "EXTERNAL_STAGE"),
                Sequence("ENABLED", Ref("EqualsSegment"), Ref("BooleanLiteralGrammar")),
                OneOf(
                    Ref("S3StorageIntegrationParameters"),
                    Ref("GCSStorageIntegrationParameters"),
                    Ref("AzureStorageIntegrationParameters"),
                ),
                Sequence(
                    "STORAGE_ALLOWED_LOCATIONS",
                    Ref("EqualsSegment"),
                    OneOf(
                        Bracketed(
                            Delimited(
                                OneOf(
                                    Ref("S3Path"),
                                    Ref("GCSPath"),
                                    Ref("AzureBlobStoragePath"),
                                )
                            )
                        ),
                        Bracketed(
                            Ref("QuotedStarSegment"),
                        ),
                    ),
                ),
                Sequence(
                    "STORAGE_BLOCKED_LOCATIONS",
                    Ref("EqualsSegment"),
                    Bracketed(
                        Delimited(
                            OneOf(
                                Ref("S3Path"),
                                Ref("GCSPath"),
                                Ref("AzureBlobStoragePath"),
                            )
                        )
                    ),
                ),
                Sequence(
                    "COMMENT",
                    Ref("EqualsSegment"),
                    Ref("QuotedLiteralSegment"),
                ),
            ),
            optional=True,
        ),
```

So I'm expecting that to be:

```
AnySetOf(
  <x>,
  <y>,
  <z>,
  optional=True,
)

```

Which I think is what you're suggesting?

**Comment by jmc-bbk** [comment]
Hold on - maybe I'm really misunderstanding - let me try remove the first `Sequence()`...

File: src/sqlfluff/dialects/dialect_snowflake.py
Comment (tunetheweb) [reply src/sqlfluff/dialects/dialect_snowflake.py:2580]: Ah yes. This was me IS needed!

**Comment by tunetheweb** [comment]
I think you’re good now actually.

**Review by tunetheweb — APPROVED** [review]
I’m happy to merge this now if you are?

**Comment by jmc-bbk** [comment]
I'm going to try and remove the `Sequence()` - will be 2 minutes! Want to test on my local before pushing :)

**Comment by jmc-bbk** [comment]
I was speaking to @WittierDinosaur and he pointed out that `AnySetOf()` is 0 or 1. In this case, I had it wrapped in `Sequence()` which expects at least one thing to be returned, and that's why the `optional=True` here was previously necessary.

Now I've unwrapped it, hopefully the tests pass - and if you agree, it should be ready for merge!

I'm definitely keen to take a crack at decomposing this class though. As with the above knowledge, the `CREATE PIPE` part below is definitely suspect (an optional sequence full of optional sequences!).

**Comment by tunetheweb** [comment]
We good to merge this now @jmc-bbk ?

**Comment by jmc-bbk** [comment]
Yep! LGTM.

**Pull request merged**
